### ISSUE TITLE
Memory leak from map-function

### ISSUE THREAD

**Issue by HJSteinort** [Slimmer-AI/mpire#73]
# Environment data
OS and version: SIMATIC Industrial OS 2.2 (based on Debian 10 (buster))
Python version: 3.7.3
Using VS Code: 1.76.0
Python extension: v2023.4.0

# Issue
I have a python module with a function that runs an infinite loop. Within this function before entering the loop I create `WorkerPool` with the mpire library.

The code looks something like this

```
def workerPoolFunction():
    mpirePool = mpire.WorkerPool(n_jobs=4)
    mpirePool.set_shared_objects(nonPicklableObject)
    mpirePool.set_keep_alive(True)

    while(True):
        results = mpirePool.map(workerFunction)
```

When I ran the function for a longterm test over the weekend the code stopped due to a memory allocation error after all the RAM got used up.
By using `psutil.Process.memory_info()` I registered that the leak is caused by `mpirePool.map()`. It is very small (4 to 100 Bytes per loop) so it was not noticable during the initial tests.

# Tests performed
I modified the code to **not** use a `shared_object` and that the `workerFunction` returns a single float variable. The leak persisted. To verify that it really is caused by the `mpirePool.map()` function I replaced it with the standard `multiprocessing.Pool.map()` function - which showed no leak...

# Limitations
Due to the nature of my application I have to exclude some solutions:

- the workers have to be kept alive because the endless loop is rather time critical and reviving the workers each loop is too time consuming
- I have to share a dictionary/list of static, non picklable objects (in my case keras-models) with all workers. Loading them from within the worker is not feasable

**Comment by sybrenjansen** [comment]
Could you perhaps check if the memory leak still persists with the latest dev branch #70 (`dev-apply`)? 

A lot will change in this PR, so I'd might already be resolved. Plan is to release it within a few weeks.

**Comment by HJSteinort** [comment]
Unfortunately the leak persits even with the latest commit of the dev branch

**Comment by sybrenjansen** [comment]
I've tried to recreate your issue with the following code:

```
In [1]: import mpire
detime: 533 ms (started: 2023-03-06 12:17:20 +01:00)

In [2]: def workerFunction(sh, x):
   ...:     y = 0
   ...:     for i in range(x):
   ...:         y += i
   ...:         y ^= 2
   ...:         y *= i
   ...:     return y
   ...: 
   ...: def workerPoolFunction():
   ...:     mpirePool = mpire.WorkerPool(n_jobs=4)
   ...:     mpirePool.set_shared_objects({1: 2, 3: 4})
   ...:     mpirePool.set_keep_alive(True)
   ...: 
   ...:     while(True):
   ...:         results = mpirePool.map(workerFunction, range(100))
```
And in another function I put tabs on `psutil.Process.memory_info()`.

When I use Python 3.7 I indeed see memory creeping up now and then.

When I switch to 3.8 or 3.9 the memory usage is stable. 

This suggests it's also not an mpire thing per se. It might be in the implementation of some of the synchronization primitives I use that vanilla multiprocessing doesn't use. 

If it's not too much trouble, I'd suggest to switch to a newer Python version, or at some intervals reset the pool:

```
   ...:     n = 0
   ...:     while(True):
   ...:         results = mpirePool.map(workerFunction, range(100))
   ...:         n += 1
   ...:         if n % 100:
   ...:             mpirePool.stop_and_join()
``` 

Notice that you don't need to revive the workers each loop, but can do so every ``x`` loops

**Comment by HJSteinort** [comment]
Thank you for looking into this issue!

Getting a higher Python version running on my OS comes with it's own errors that I'm currently trying to figure out.
I will come back to you after I can confirm that upgrading Python stopped the leak.

**Comment by sybrenjansen** [comment]
Hmm.. I tried again with python 3.7 and again saw the leak again as expected. But then, I added

```
import gc

gc.collect()
```
to the end of `workerFunction` such that at the end of a task the worker cleans up any unnecessary resources it still holds (garbage collection).

This made the leak go away as well. Perhaps you can try that.

If that still doesn't work, then it must be very specific to what you're running. In that case, can you try to make a minimal working example that I can run so I can test it here locally as well? Otherwise it will be really hard to debug this.

**Comment by HJSteinort** [comment]
While creating a minimal working example I recognized, that the strange behaviour might be caused by my system.

When debugging the script with VS Code I used python 3.10.10 as interpreter from within a virtual environment. This works but still shows the memory leak as if I didn't upgrade python.
Creating the debugFile for you I did the same and still saw the memory leak in the same pace.
The strange thing is if I activate the virtual environment in my terminal and execute the script directly from within this environment like 

`(debugEnv) user: $ python debugFile.py`

The leak vanishes and the memory utilization is stable after the loop being executed a couple of times.

In short your reccomended solution of upgrading python to 3.8+ worked and the root cause of my leak seems to lie within my system.

**Comment by sybrenjansen** [comment]
Is https://stackoverflow.com/questions/75603646/mpire-workerpool-causes-memory-leak your issue by the way?

**Comment by HJSteinort** [comment]
Yes, that was my issue. I closed it with a link to this discussion 😸
